plotly/graph_objs/graph_objs.py

```diff
@@ -297,67 +348,175 @@ def force_clean(self, caller=True):
-    subclasses and containers so long as all plot objects are instantiated
-    as a subclass of PlotlyDict. Each PlotlyDict should be instantiated
-    with a `kind` keyword argument. This defines the special _info
-    dictionary for the object.
-
-    Any available methods that hold for a dict hold for a PlotlyDict.
+class PlotlyDict(dict, PlotlyBase):
+    """
+    Base class for dict-like Plotly objects.
 
     """
+    _name = None
+    _attributes = set()
+    _parent_key = None
 
     def __init__(self, *args, **kwargs):
-        class_name = self.__class__.__name__
+        if self._name is None:
+            raise exceptions.PlotlyError(
+                "PlotlyDict is a base class. It's shouldn't be instantiated."
+            )
 
-        for key in kwargs:
-            if utils.is_source_key(key):
-                kwargs[key] = self._assign_id_to_src(key, kwargs[key])
+        _raise = kwargs.pop('_raise', True)
 
-        super(PlotlyDict, self).__init__(*args, **kwargs)
-        if issubclass(NAME_TO_CLASS[class_name], PlotlyTrace):
-            if (class_name != 'PlotlyTrace') and (class_name != 'Trace'):
-                self['type'] = NAME_TO_KEY[class_name]
-        self.validate()
-        if self.__class__.__name__ == 'PlotlyDict':
-            warnings.warn("\nThe PlotlyDict class is a base class of "
-                          "dictionary-like graph_objs.\nIt is not meant to be "
-                          "a user interface.")
+        super(PlotlyDict, self).__init__()
 
-    def __setitem__(self, key, value):
+        if self._name in graph_reference.TRACE_NAMES:
+            self['type'] = self._name
 
-        if key in ('xsrc', 'ysrc'):
-            value = self._assign_id_to_src(key, value)
+        # force key-value pairs to go through validation
+        d = {key: val for key, val in dict(*args, **kwargs).items()}
+        for key, val in d.items():
+            try:
+                self.__setitem__(key, val, _raise=_raise)
+            except exceptions.PlotlyGraphObjectError as err:
+                err.prepare()
+                raise
 
-        return super(PlotlyDict, self).__setitem__(key, value)
+    def __dir__(self):
+        attrs = self.__dict__.keys()
+        attrs += [attr for attr in dir(dict()) if attr not in attrs]
+        return sorted(self._attributes) + attrs
+
+    def __getitem__(self, key):
+        if key not in self:
+            self.__missing__(key)
+        return super(PlotlyDict, self).__getitem__(key)
+
+    def __setattr__(self, key, value):
+        self.__setitem__(key, value)
+
+    def __setitem__(self, key, value, _raise=True):
+
+        if not isinstance(key, six.string_types):
+            if _raise:
+                raise TypeError('Key must be string, not {}'.format(type(key)))
+            return
+
+        if key.endswith('src') and key in self._attributes:
+            value = graph_objs_tools.assign_id_to_src(key, value)
```

Comment:
@chriddyp ? i believe this is something he added a while back, just maintaining.